Applies hunks in single patch call

lisroach opened the pull request:
This should solve[ issue 19](https://github.com/facebookincubator/Bowler/issues/19) by adding all the hunks at once. 
I stripped out the name of the file (the --- filename +++ filename) from the top of each hunk, and created one giant bytestring with the hunks all combined. 

`make lint test` passes.

ambv commented on bowler/tool.py:
```diff
@@ -281,14 +283,19 @@ def process_hunks(self, filename: Filename, hunks: List[Hunk]) -> None:
+                need_patch = True
+                accepted_hunks += "\n".join(hunk[2:]) + "\n"
+
+        if need_patch:
```
nit: Since this is the end of the function, you can do "`if not need_patch: return`" instead to save an indentation level and make it clearer to the reader that this situation is already handled.

Better yet, you could skip the need for `need_patch` altogether if you start with an empty `accepted_hunks` string and only add the header if `accepted_hunks` is non-empty after the for-loop.

ambv commented on bowler/tool.py:
```diff
@@ -281,14 +283,19 @@ def process_hunks(self, filename: Filename, hunks: List[Hunk]) -> None:
+                accepted_hunks += "\n".join(hunk[2:]) + "\n"
+
+        if need_patch:
+            args = ["patch", "-u", filename]
```
Would be cool to actually use those args in the call below. You could say: `sh.patch(*args[1:], _in=...)`.

ambv commented on bowler/tool.py:
```diff
@@ -281,14 +283,19 @@ def process_hunks(self, filename: Filename, hunks: List[Hunk]) -> None:
+                )  # type: ignore
+            except sh.ErrorReturnCode:
+                log.exception("failed to apply patch hunk")
+                return  # TODO: better response?
```
Agreed, a better response would be great. A good way to find out what response would be helpful is to make a test that exercises this path :-)

jreese commented on bowler/tool.py:
```diff
@@ -286,14 +287,25 @@ def process_hunks(self, filename: Filename, hunks: List[Hunk]) -> None:
+                accepted_hunks += "\n".join(hunk[2:]) + "\n"
+
+        if accepted_hunks:
+            accepted_hunks = f"--- {filename}\n+++ {filename}\n" + accepted_hunks
```
I believe it would be faster to just include `accepted_hunks` in the f-string, rather than rendering the f-string and then concatenating.

jreese commented on bowler/tool.py:
```diff
@@ -286,14 +287,25 @@ def process_hunks(self, filename: Filename, hunks: List[Hunk]) -> None:
+                        log.exception(f"hunks failed to apply, rejects saved to{err}")
+                        return
+                log.exception(f"failed to apply patch hunk: {err}")
+                return
```
unnecessary return?

lisroach commented:
Thanks for pointing that out @jreese, I've refactored a little and added tests to handle when users cancel after some hunks have already been applied. Let me know if you think this should work better
